q2_diversity/_alpha/_visualizer.py

```diff
@@ -193,3 +201,147 @@ def alpha_correlation(output_dir: str,
                     os.path.join(output_dir, 'dist'))
+
+
+def _reindex_with_metadata(category, categories, merged):
+    merged.set_index(category, inplace=True)
+    merged.sort_index(axis=0, ascending=True, inplace=True)
+    merged = merged.groupby(level=[category])
+    counts = merged.count()
+    counts.drop(categories, axis=1, inplace=True, level=0)
+    sum_ = merged.sum()
+    return sum_, counts
+
+
+def _compute_summary(data, id_label, counts=None):
+    perc = [0.02, 0.09, 0.25, 0.5, 0.75, 0.91, 0.98]
+    describer = functools.partial(pd.DataFrame.describe, percentiles=perc)
+    summary_df = data.stack(level=0)
+    summary_df = summary_df.apply(describer, axis=1)
+    summary_df.drop(['std', 'mean'], axis=1, inplace=True)
+    if counts is not None:
+        summary_df.drop('count', axis=1, inplace=True)
+        stacked_counts = counts.stack(level=0)
+        # There will always be at least one iteration, so we grab the first
+        stacked_counts = stacked_counts[[1]]
+        stacked_counts.rename(columns={1: 'count'}, inplace=True)
+        summary_df = summary_df.join(stacked_counts, how='inner')
+    else:
+        # Reset count (this should always be one if we weren't explicitly
+        # passed counts)
+        summary_df['count'] = 1
+    summary_df = summary_df.reset_index()
+    summary_df.rename(columns={'level_0': id_label}, inplace=True)
+    return summary_df
+
+
+def _beta_rarefaction_jsonp(output_dir, filename, metric, data, category):
+    with open(os.path.join(output_dir, filename), 'w') as fh:
+        fh.write("load_data('%s', '%s'," % (metric, category))
+        data.to_json(fh, orient='split')
+        fh.write(");")
+
+
+def _compute_rarefaction_data(feature_table, min_depth, max_depth, steps,
+                              iterations, phylogeny, metrics):
+    depth_range = np.linspace(min_depth, max_depth, num=steps, dtype=int)
+    iter_range = range(1, iterations + 1)
+
+    rows = feature_table.ids(axis='sample')
+    cols = pd.MultiIndex.from_product([list(depth_range), list(iter_range)],
+                                      names=['depth', 'iter'])
+    data = {k: pd.DataFrame(np.NaN, index=rows, columns=cols)
+            for k in metrics}
+
+    for d, i in itertools.product(depth_range, iter_range):
+        rt = rarefy(feature_table, d)
+        for m in metrics:
+            if m in phylogenetic_metrics():
+                vector = alpha_phylogenetic(table=rt, metric=m,
+                                            phylogeny=phylogeny)
+            else:
+                vector = alpha(table=rt, metric=m)
+            data[m][(d, i)] = vector
+    return data
+
+
+def alpha_rarefaction(output_dir: str, table: biom.Table, max_depth: int,
+                      phylogeny: skbio.TreeNode=None, metric: str=None,
+                      metadata: qiime2.Metadata=None, min_depth: int=1,
+                      steps: int=10, iterations: int=10) -> None:
+    if steps < 2:
```

Comment:
Thanks guys --- the secondary validation was intended for folks who skip q2 entirely and use the function outright. I am good either way (TODO: look into exposing validation hooks for that use case).